### PR TITLE
feat: ntfy alerts for monitor logs. Closes #664

### LINKED ISSUE #664: Altert via ntfy on errors
Currently there is an option to receive alerts via Slack in the MonitorLogs job. It would be great to also have the option to receive alters via ntfy.

### PR DESCRIPTION
**Pull request by HARSHVARANDANI**
# Description

This feature extends the existing Slack-based alerting in MonitorLogs by also publishing notifications to a configured ntfy topic when errors are detected in monitored logs.

## Related issues
Please add related issues.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality).

# Checklist

- [ x] I have read and understood the rules about [how to Contribute](https://intelowlproject.github.io/docs/GreedyBear/Contribute/) to this project.
- [x] The pull request is for the branch `develop`.
- [x] I have added documentation of the new features.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowlproject.github.io/docs/GreedyBear/Contribute/#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] I have added tests for the feature/bug I solved. All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/docs/blob/main/docs/GreedyBear/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.
  
### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

### CODE REVIEW

File: docker/env_file_template [hunk 1]
```diff
@@ -35,6 +35,8 @@ ELASTIC_ENDPOINT=
 SLACK_TOKEN=
 DEFAULT_SLACK_CHANNEL=
 
+NTFY_URL=
+
```
Comment (regulartim) [inline docker/env_file_template:41]: I would prefer to only have the nfty topic as a variable, not the whole URL. But I am open to a discussion about that. Do you think it is better to have the whole URL as a variable?

File: greedybear/cronjobs/monitor_logs.py [hunk 2]
```diff
@@ -47,5 +48,6 @@ def run(self):
                 message = f"found errors in log file {log_file}"
                 self.log.warning(message)
                 send_message(message)
```
Comment (regulartim) [inline greedybear/cronjobs/monitor_logs.py]: Do you think it would be good to rename this function to `send_slack_message` or something?

File: greedybear/migrations/0001_initial.py [hunk 3]
```diff
@@ -46,9 +46,7 @@ class Migration(migrations.Migration):
                 ("last_seen", models.DateTimeField(default=datetime.datetime.utcnow)),
                 (
                     "days_seen",
-                    django.contrib.postgres.fields.ArrayField(
-                        base_field=models.DateField(), blank=True, size=None
-                    ),
+                    django.contrib.postgres.fields.ArrayField(base_field=models.DateField(), blank=True, size=None),
```
Comment (regulartim) [inline greedybear/migrations/0001_initial.py]: Please don't include files that have nothing to do with the issue you are working on.

File: greedybear/migrations/0004_alter_id_field.py [hunk 4]
```diff
@@ -13,22 +13,16 @@ class Migration(migrations.Migration):
         migrations.AlterField(
             model_name="ioc",
             name="id",
-            field=models.BigAutoField(
-                auto_created=True, primary_key=True, serialize=False, verbose_name="ID"
-            ),
+            field=models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name="ID"),
```
Comment (regulartim) [inline greedybear/migrations/0004_alter_id_field.py]: Please don't include files that have nothing to do with the issue you are working on.

File: greedybear/migrations/0005_clients.py [hunk 5]
```diff
@@ -8,9 +8,7 @@ def create_default_clients(apps, schema_editor):
     # version than this migration expects. We use the historical version.
     Client = apps.get_model("durin", "Client")
     # for pygreedybear, custom token_ttl
-    Client.objects.update_or_create(
-        name="pygreedybear", token_ttl=timedelta(weeks=4 * 12 * 10)
-    )
+    Client.objects.update_or_create(name="pygreedybear", token_ttl=timedelta(weeks=4 * 12 * 10))
```
Comment (regulartim) [inline greedybear/migrations/0005_clients.py]: Please don't include files that have nothing to do with the issue you are working on.

File: greedybear/migrations/0010_alter_ioc_related_ioc.py [hunk 6]
```diff
@@ -6,13 +6,13 @@
 class Migration(migrations.Migration):
 
     dependencies = [
-        ('greedybear', '0009_alter_ioc_general_field'),
+        ("greedybear", "0009_alter_ioc_general_field"),
```
Comment (regulartim) [inline greedybear/migrations/0010_alter_ioc_related_ioc.py:8]: Please don't include files that have nothing to do with the issue you are working on.

File: requirements/project-requirements.txt [hunk 7]
```diff
@@ -21,3 +21,5 @@ pandas==2.3.3
 scikit-learn==1.8.0
 numpy==2.4.0
 datasketch==1.8.0
+
+requests>=2.31,<3
```
Comment (regulartim) [inline requirements/project-requirements.txt]: Do we really need that import? `requests` is part of the python core lib, right?

**Review by regulartim — CHANGES_REQUESTED** [review]
Nice!

**Comment by regulartim** [comment]
Also: do you think you can write a test for that?

File: docker/env_file_template (hunk 1 above)
Comment (HARSHVARANDANI) [reply docker/env_file_template:41]: @regulartim  My apprach was this would be easier to integrate self hosted ntfy since it might have a localhost or maybe an http url so only one url would need to be given in the env directly and it would be less of a hassle. What would you suggest?

File: greedybear/cronjobs/monitor_logs.py (hunk 2 above)
Comment (HARSHVARANDANI) [reply greedybear/cronjobs/monitor_logs.py]: Yeah I guess it would make the code more readable will do it :+1:

File: greedybear/migrations/0001_initial.py (hunk 3 above)
Comment (HARSHVARANDANI) [reply greedybear/migrations/0001_initial.py]: Sorry about that I guess these were added by the pre-commit changes, will fix this :+1:

File: requirements/project-requirements.txt (hunk 7 above)
Comment (HARSHVARANDANI) [reply requirements/project-requirements.txt]: I don't think so that requests is pre installed with the python interpreter, it needs to be installed separately as far as I know but I might be wrong.
image

File: docker/env_file_template (hunk 1 above)
Comment (regulartim) [reply docker/env_file_template:41]: Good point! Let's keep the URL as a variable.

File: requirements/project-requirements.txt (hunk 7 above)
Comment (regulartim) [reply requirements/project-requirements.txt]: You are right, sorry for the mistake.

**Comment by regulartim** [comment]
And another thing: would you add headers to the request, such that the alerts have a prettier format? Suggestion (feel free to improve):
```python
headers={
        "Title": "GreedyBear error",
        "Tags": "warning"
    })
```

**Review by regulartim — CHANGES_REQUESTED** [review]
Good job, I like it! 👍 
Could you please add a set of test cases for `send_ntfy_message`?

**Comment by HARSHVARANDANI** [comment]
> Good job, I like it! 👍 Could you please add a set of test cases for `send_ntfy_message`?

Just to be sure, do you mean add a testing python script like the ones in the test folder?

**Comment by regulartim** [comment]
> Just to be sure, do you mean add a testing python script like the ones in the test folder?

Yes, exactly. It should cover happy paths, failing paths, edge cases. Just take a look at a few existing tests to get an impression.

File: requirements/project-requirements.txt (hunk 7 above)
Comment (regulartim) [inline requirements/project-requirements.txt]: I've checked the dependency tree and `requests` is already installed as a transitive dependency for `certego_saas`. So to avoid possible conflicts, we should remove this line. Do you agree?

File: docker/env_file_template [hunk 8]
```diff
@@ -35,6 +35,8 @@ ELASTIC_ENDPOINT=
 SLACK_TOKEN=
 DEFAULT_SLACK_CHANNEL=
 
+NTFY_URL=
```
Comment (regulartim) [inline docker/env_file_template:38]: Maybe some form of explanation for the variable would be nice. Something like:
```
# Set to your ntfy topic URL to receive push notifications
# Example: https://ntfy.sh/your-topic-name
```

**Review by regulartim — CHANGES_REQUESTED** [review]
We are getting close! :) Just a few minor things left. Also: please fix the formatting such that the CI is happy.

File: requirements/project-requirements.txt (hunk 7 above)
Comment (HARSHVARANDANI) [reply requirements/project-requirements.txt]: Alright if thats the case then we should remove it.

Comment (HARSHVARANDANI) [reply requirements/project-requirements.txt]: Done :+1:

File: docker/env_file_template (hunk 8 above)
Comment (HARSHVARANDANI) [reply docker/env_file_template:38]: Done :+1:

**Review by regulartim — APPROVED** [review]

**Comment by regulartim** [comment]
@HARSHVARANDANI Fix the formatting such that the CI is happy, then we can merge.

**Comment by HARSHVARANDANI** [comment]
> Fix the formatting such that the CI is happy, then we can merge.

@regulartim I have already used pre-commit before every commit to check for the formatting Black, Flake and Isort, is there anything else there is supposed to be done for the formatting?

**Comment by regulartim** [comment]
Ok, that's strange. Maybe this is because of the ongoing migration to Ruff. I'll check and get back to you.

File: tests/test_ntfy.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+TEST_LOGGING = {
+    "version": 1,
+    "disable_existing_loggers": True,
+}
+
+
+@override_settings(LOGGING=TEST_LOGGING)
+class SendNtfyMessageTests(SimpleTestCase):
+
+    @override_settings(NTFY_URL="https://ntfy.sh/greedybear")
```
Comment (regulartim) [inline tests/test_ntfy.py:15]: I fond the issue, this is indeed because of the recent integration of ruff for linting. Please remove the empty line (14) here. 

For future PRs, please use ruff as described in our [contribution guidelines](https://intelowlproject.github.io/docs/GreedyBear/Contribute/).

Comment (HARSHVARANDANI) [reply tests/test_ntfy.py:15]: Sure, will do 👍

**Comment by regulartim** [comment]
Now ruff is unhappy with the import ordering and I don't know why. I cloned @HARSHVARANDANI 's repo and ran ruff using our `.ruff.toml` from the develop branch. It did not complain about anything. Can you please help @mlodic @opbot-xd ?

**Comment by mlodic** [comment]
I tried it locally and I correctly replicate the error reported by the CI. Ruff is working fine in my env. How did you install pre-commit/ruff in your env?

So yes, the error must be fixed

**Comment by opbot-xd** [comment]
image
I ran the ruff check command it is showing me the same error
and on applying fix command the diff is only
the fix is to add a line after 

```python
from unittest.mock import MagicMock, patch

from django.test import SimpleTestCase, override_settings

from greedybear.ntfy import send_ntfy_message

TEST_LOGGING = {
 ```
 
image

after then on running the check again, all checks passed
so try this once
since i did not have the push access so was not able to push my code changes

**Comment by mlodic** [comment]
ah you know what? It could be that you executed the ruff linter instead of the ruff formatter. Those are 2 different commands. @regulartim 
To be sure to execute all of them, please run `pre-commit run ruff --all-files -c .github/.pre-commit-config.yaml`

**Comment by regulartim** [comment]
Sorry, my bad. I ran `ruff format` instead of `ruff check`. Thanks @mlodic @opbot-xd !

**Comment by regulartim** [comment]
Now the CI is happy! Thanks for your work @HARSHVARANDANI, I 'll merge now.
